Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch

```diff
@@ -0,0 +1,177 @@
++        Packet offhand = new PacketPlayOutEntityEquipment(getEntityId(), EnumItemSlot.OFFHAND, CraftItemStack.asNMSCopy(getInventory().getItemInOffHand()));
++
++        Packet helm = new PacketPlayOutEntityEquipment(getEntityId(), EnumItemSlot.HEAD, CraftItemStack.asNMSCopy(getInventory().getHelmet()));
++
++        Packet chest = new PacketPlayOutEntityEquipment(getEntityId(), EnumItemSlot.CHEST, CraftItemStack.asNMSCopy(getInventory().getChestplate()));
++
++        Packet legs = new PacketPlayOutEntityEquipment(getEntityId(), EnumItemSlot.LEGS, CraftItemStack.asNMSCopy(getInventory().getLeggings()));
++
++        Packet feet = new PacketPlayOutEntityEquipment(getEntityId(), EnumItemSlot.FEET, CraftItemStack.asNMSCopy(getInventory().getBoots()));
++
++        Packet slot = new PacketPlayOutHeldItemSlot(getInventory().getHeldItemSlot());
++
++        for (Player pOnline : Bukkit.getOnlinePlayers()) {
++            EntityPlayer handle = ((CraftPlayer) pOnline).getHandle();
++            PlayerConnection playerCon = handle.playerConnection;
++            if (pOnline.equals(this)) {
++                reregisterPlayer(handle);
++
++                //Respawn the player then update their position and selected slot
++                handle.playerConnection.sendPacket(respawn);
++                handle.updateAbilities();
++                handle.playerConnection.sendPacket(pos);
++                handle.playerConnection.sendPacket(slot);
++
++                ((CraftPlayer) pOnline).updateScaledHealth();
++                pOnline.updateInventory();
++
++                handle.triggerHealthUpdate();
++
++                if (pOnline.isOp()) {
++                    pOnline.setOp(false);
++                    pOnline.setOp(true);
++                }
++
++                continue;
++            }
++            if (pOnline.getWorld().equals(getWorld()) && pOnline.canSee(this) && isOnline()) {
++                PacketPlayOutEntityDestroy removeEntity = new PacketPlayOutEntityDestroy(new int[]{getEntityId()});
++                PacketPlayOutNamedEntitySpawn addNamed = new PacketPlayOutNamedEntitySpawn(getHandle());
++
++                //Remove and reregister the player
++                handle.playerConnection.sendPacket(removeEntity);
++                ((CraftPlayer) pOnline).reregisterPlayer(this.getHandle());
++                handle.playerConnection.sendPacket(addNamed);
++
++
++                //Send hand items
++                handle.playerConnection.sendPacket(mainhand);
++                handle.playerConnection.sendPacket(offhand);
++
++
++                //Send armor
++                handle.playerConnection.sendPacket(helm);
++                handle.playerConnection.sendPacket(chest);
++                handle.playerConnection.sendPacket(legs);
++                handle.playerConnection.sendPacket(feet);
++
++            } else {
++                //Just send player update
++                ((CraftPlayer) pOnline).reregisterPlayer(this.getHandle());
```

Comment:
This was actually a major issue. We don't want to touch anything for the player if the the receiving player cant see the source player

Reply:
pOnline.canSee(this)